rl_equation_solver/utilities/utilities.py

```diff
@@ -0,0 +1,316 @@
+        link_list.append({"source": parent.id, "target": node.id})
+
+    else:
+        node = Node(str(type(expr).__name__), Id.get())
+        node_list.append({"id": node.id, "name": node.name})
+        link_list.append({"source": parent.id, "target": node.id})
+        for arg in expr.args:
+            graph_walk(node, arg, node_list, link_list)
+
+
+def pad_array(arr, length):
+    """
+    Pad array with zeros according the given length
+    """
+    if len(arr) < length:
+        padded_arr = np.zeros(length)
+        padded_arr[:len(arr)] = arr
+        return padded_arr
+    else:
+        return arr
+
+
+def to_vec(expr, feature_dict, state_dim=4096):
+    """
+    Get state vector for given expression
+
+    Parameters
+    ----------
+    expr : str
+        State string representation
+    feature_dict : dict
+        Dictionary mapping feature names to values
+    state_dim : int
+        Max length of state vector
+
+    Returns
+    -------
+    np.ndarray
+        State vector array
+    """
+    graph = get_json_graph(expr)
+    node_features = get_node_features(graph, feature_dict)
+    node_features = pad_array(node_features, int(0.25 * state_dim))
+    edge_vector = nx.to_numpy_array(graph).flatten()
+    edge_vector = pad_array(edge_vector, int(0.75 * state_dim))
+    state_vec = np.concatenate([node_features, edge_vector])
+
+    return state_vec
+
+
+def get_json_graph(expr):
+    """
+    Make a graph plot of the internal representation of SymPy expression. Don't
+    add meta data yet.
+
+    Parameters
+    ----------
+    expr : str
+        State string representation
+
+    Returns
+    -------
+    networkx.Graph
+    """
+    node_list = []
+    link_list = []
+
+    graph_walk(Node("Root", -1), expr, node_list, link_list)
+
+    # Create the graph from the lists of nodes and links:
+    graph_json = {"nodes": node_list, "links": link_list}
+    node_labels = {node['id']: node['name'] for node
+                   in graph_json['nodes']}
+    for n in graph_json['nodes']:
+        del n['name']
+    graph = json_graph.node_link_graph(graph_json, directed=True,
+                                       multigraph=False)
+    for node in graph.nodes:
+        graph.nodes[node]['name'] = node_labels.get(node, 'Root')
+    graph.remove_node(-1)
+    return graph
+
+
+def to_graph(expr, feature_dict):
+    """
+    Make a graph plot of the internal representation of SymPy expression.
+
+    Parameters
+    ----------
+    expr : str
+        State string representation
+    feature_dict : dict
+        Dictionary mapping feature names to values
+
+    Returns
+    -------
+    networkx.Graph
+    """
+    graph = get_json_graph(expr)
+    node_features = get_node_features(graph, feature_dict)
+    for i, node in enumerate(list(graph.nodes)):
+        graph.nodes[node]['x'] = node_features[i]
+    return graph
+
+
+def get_node_features(graph, feature_dict):
+    """Get node features from feature dictionary"""
```

Comment:
Example of node features: we map the operations and terms to integeters: {add: 0, subt: 1, .. }